### PR TITLE
calculate fees

### LINKED ISSUE #2230: Calculate fees for every transaction
**Is your feature request related to a problem? Please describe.**

As a developer before implementing #2008
I want to know the fees needed for a transaction
So that I can implement the view layer.

**Describe the solution you'd like**

On every transaction calculate the fees (without displaying them  yet).

**Describe alternatives you've considered**

**Additional context**

AC:
- calculate the fees for a transaction before a transaction
  - simulate the gas with a request to stargate
  - calculate fees in atoms
- print the calculated fees to the console

### PR DESCRIPTION
**Pull request by fedekunze**
Closes #2230 

**Description:**

Screen Shot 2019-03-14 at 10 59 58 PM

Thank you! 🚀

---

For contributor:

- [ ] Added entries in `PENDING.md` with issue # and GitHub username
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by faboweb** [comment]
Isn't the issue about calculating fees?

**Comment by fedekunze** [comment]
It’s not the UI, just adding the fees to the total amount on transaction

**Comment by faboweb** [comment]
Should we add the fees to the total amount or should we rather show the fees or "shipment", as it would be in e-commerce, extra?

**Comment by fedekunze** [comment]
> Should we add the fees to the total amount or should we rather show the fees or "shipment", as it would be in e-commerce, extra?

So there're 2 approaches:

1. **USA**: The price doesn't include the tax (fees in our case)
2. **EU**: The price includes the tax (fees in our case)

on this PR I has taking the 2nd one for the tx total amount. Not sure what you guys think ?

cc: @jbibla

**Comment by faboweb** [comment]
I wouldn't make the comparison "fees" = "taxs" but the comparison "fees" = "shippment".

**Comment by fedekunze** [comment]
we need your wisdom here, oh UX almighty @jbibla

File: test/unit/specs/components/transactions/__snapshots__/LiStakeTransaction.spec.js.snap [hunk 1]
```diff
@@ -78,8 +78,15 @@ exports[`LiStakeTransaction edit validator 1`] = `
...
+    <!---->
+  </div>
+   
+  <div>
+    
+      Moniker: 
+      
     <router-link-stub
-      to="/validator/cosmosvaloper1qecshyc40kshszkwrtscgmsdd8tz3n4hrj9yf2"
+      to="/validator/undefined"
```
Comment (faboweb) [inline test/unit/specs/components/transactions/__snapshots__/LiStakeTransaction.spec.js.snap]: warning

**Comment by jbibla** [comment]
> we need your wisdom here, oh UX almighty @jbibla

i am but a simple man.

File: app/src/renderer/vuex/modules/send.js [hunk 2]
```diff
@@ -71,14 +81,25 @@ export default ({ node }) => {
 
       await dispatch(`queryWalletBalances`) // the nonce was getting out of sync, this is to force a sync
 
+      const gasPrices = [
+        {
+          amount: `0.025`, // recommended on Cosmos Docs
+          denom: `stake` //TODO: this shouldn't be hardcoded
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/send.js]: isn't this going to be an issue immediately?

File: app/src/renderer/vuex/modules/send.js [hunk 3]
```diff
@@ -97,18 +118,30 @@ export default ({ node }) => {
...
+      // simulation to get the estimated gas
+      let request = actions.apiRequest(type, to, pathParameter, args)
+      const simulationRes = await request.catch(handleSDKError)
+      const adjustedGas = Number(simulationRes.gas_estimate) * gasAdjustment
+      const estimatedFeesLow =
+        Number(simulationRes.gas_estimate) * Number(gasPrices[0].amount) * 1e-7
+      const estimatedFeesHigh =
+        Number(adjustedGas) * Number(gasPrices[0].amount) * 1e-7
+
+      console.log(
```
Comment (jbibla) [inline app/src/renderer/vuex/modules/send.js]: console.log? is this temporary until the info is displayed in the UI?

Comment (fedekunze) [reply app/src/renderer/vuex/modules/send.js]: yup, as @faboweb suggested

Comment (fedekunze) [reply app/src/renderer/vuex/modules/send.js]: I can work on the action modal as soon as we merge this

File: app/src/renderer/vuex/modules/send.js (hunk 2 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/send.js]: ohh true, good catch !

File: app/src/renderer/components/transactions/LiAnyTransaction.vue [hunk 4]
```diff
@@ -88,6 +93,12 @@ export default {
     type() {
       return this.transaction.tx.value.msg[0].type
     },
+    fees() {
+      return (
+        this.transaction.tx.value.fee.amount &&
```
Comment (sabau) [inline app/src/renderer/components/transactions/LiAnyTransaction.vue]: May happen that `this.transaction.tx.value.fee` is `undefined`?

File: app/src/renderer/components/transactions/LiBankTransaction.vue [hunk 5]
```diff
@@ -6,21 +6,25 @@
   >
     <template v-if="address === ''">
       <div slot="caption">
-        Sent&nbsp;<b>{{ full(atoms(coins.amount)) }}</b>
-        <span>&nbsp;{{ coins.denom.toUpperCase() }}</span>
+        Sent<b>{{ full(atoms(coins.amount)) }}</b>
```
Comment (sabau) [inline app/src/renderer/components/transactions/LiBankTransaction.vue]: I would keep a space between `Sent` and the amount

File: app/src/renderer/vuex/modules/send.js [hunk 6]
```diff
@@ -97,18 +118,31 @@ export default ({ node }) => {
...
-        request = node[type](args)
-      }
+      // simulation to get the estimated gas
+      let request = actions.apiRequest(type, to, pathParameter, args)
+      const simulationRes = await request.catch(handleSDKError)
+      const adjustedGas = Number(simulationRes.gas_estimate) * gasAdjustment
+      const estimatedFeesLow =
+        Number(simulationRes.gas_estimate) * Number(gasPrices[0].amount) * 1e-7
+      const estimatedFeesHigh =
+        Number(adjustedGas) * Number(gasPrices[0].amount) * 1e-7
```
Comment (sabau) [inline app/src/renderer/vuex/modules/send.js]: just double checking the 1e-7 XD

File: app/src/renderer/components/transactions/LiBankTransaction.vue (hunk 5 above)
Comment (fedekunze) [reply app/src/renderer/components/transactions/LiBankTransaction.vue]: that's a double space on the UI, @jbibla found

File: app/src/renderer/vuex/modules/send.js (hunk 6 above)
Comment (fedekunze) [reply app/src/renderer/vuex/modules/send.js]: damn true lol

Comment (fedekunze) [reply app/src/renderer/vuex/modules/send.js]: updated

File: app/src/renderer/components/transactions/LiBankTransaction.vue (hunk 5 above)
Comment (sabau) [reply app/src/renderer/components/transactions/LiBankTransaction.vue]: Funny, it seems no differences at occurs by adding or removing it

File: app/src/renderer/components/transactions/LiAnyTransaction.vue (hunk 4 above)
Comment (fedekunze) [reply app/src/renderer/components/transactions/LiAnyTransaction.vue]: yes, if you send a tx without fees it returns `null`

Comment (faboweb) [reply app/src/renderer/components/transactions/LiAnyTransaction.vue]: He meant `this.transaction.tx.value.fee`

File: app/src/renderer/vuex/modules/send.js [hunk 7]
```diff
@@ -62,6 +62,16 @@ export default ({ node }) => {
     resetSessionData({ state }) {
       state.nonce = `0`
     },
+    apiRequest(type, to, pathParameter, args) {
+      // get the generated tx by querying it from the backend
+      if (to && pathParameter) {
```
Comment (faboweb) [inline app/src/renderer/vuex/modules/send.js]: this still needs a refactor of `api.js`

Comment (jbibla) [reply app/src/renderer/vuex/modules/send.js]: what do you mean?

**Pull request merged**